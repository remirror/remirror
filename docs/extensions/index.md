---
hide_title: true
title: Overview
---

# All extensions

## Standard extensions

The following extensions are bundled together with Remirror:

**[AnnotationExtension](./annotation-extension.mdx)**<br /> Allows the annotation (or highlighting) of the content in your editor

**[BlockquoteExtension](./blockquote-extension.mdx)**<br /> Makes regions of text appear as an indented quote to ensure it stands out

**[BoldExtension](./bold-extension.mdx)**<br /> Makes the text under the cursor / or at the provided position range **bold**

**[CalloutExtension](./callout-extension.mdx)**<br /> Makes regions of text stand out on a colour background with optional emoji prefix

**[CodeBlockExtension](./code-block-extension.mdx)**<br /> Adds code blocks to the editor. This differs to the `CodeExtension`, which provides code marks on inline text.

**[CodeExtension](./code-extension.mdx)**<br /> Makes the text under the cursor / or at the provided position range `code`

**[CollaborationExtension](./collaboration-extension.mdx)**<br /> Adds collaborative functionality powered by `prosemirror-collab`

**[ColumnsExtension](./columns-extension.mdx)**<br /> Add column support to the nodes

**[DocExtension](./doc-extension.mdx)**<br /> The root node in your editor. Allows you to customise attributes and valid `content`.

**[DropCursorExtension](./drop-cursor-extension.mdx)**<br /> Shows a line indicator for where the drop target will be

**[EmbedExtension](./embed-extension.mdx)**<br /> Adds iframe-based embeds to your text editor.

**[EmojiExtension](./emoji-extension.mdx)**<br /> Adds emoji support

**[EntityReferenceExtension](./entity-reference-extension.mdx)**<br /> Allows you to reference external entities in the content of your editor.

**[EpicModeExtension](./epic-mode-extension.mdx)**<br /> Shows the editor in epic mode.

**[EventsExtension](./events-extension.mdx)**<br /> Listens to events which occur within the editor

**[FindExtension](./find-extension.mdx)** <br /> Add find and replace functionality to your editor.

**[FontFamilyExtension](./font-family-extension.mdx)**<br /> Add a font family to the selected text

**[FontSizeExtension](./font-size-extension.mdx)**<br /> Add a font size to the selected text

**[GapCursorExtension](./gap-cursor-extension.mdx)**<br /> This will capture clicks near and arrow-key-motion past places that don't have a normally selectable position nearby, and create a gap cursor selection for them.

**[HardBreakExtension](./hard-break-extension.mdx)**<br /> Inserts a `hardBreak` `<br />` tag into the editor

**[HeadingExtension](./heading-extension.mdx)**<br /> Adds headings to the editor

**[HistoryExtension](./history-extension.mdx)**<br /> Provides undo and redo commands

**[HorizontalRuleExtension](./horizontal-rule-extension.mdx)**<br /> Adds a horizontal line `<hr />` tag to the editor

**[ImageExtension](./image-extension.mdx)**<br /> Places images into the editor

**[ItalicExtension](./italic-extension.mdx)**<br /> Makes the text under the cursor / or at the provided position range _italic_

**[LinkExtension](./link-extension.mdx)**<br /> Makes the text under the cursor / or at the provided position range link to another resource

**[ListExtension](./list-extension.mdx)**<br /> Adds ordered lists, unordered lists, and checklists to the editor

**[MarkdownExtension](./markdown-extension.mdx)**<br /> Transforms the **ProseMirror** content of your editor to markdown syntax

**[MentionAtomExtension](./mention-atom-extension.mdx)**<br /> Provides mentions as atom nodes which don't support editing once being inserted into the document

**[MentionExtension](./mention-extension.mdx)**<br /> Provides mentions as nodes which support editing once being inserted into the document

**[NodeFormattingExtension](./node-formatting-extension.mdx)**<br /> Enables you to align, index and set the line-height of content in your editor

**[PlaceholderExtension](./placeholder-extension.mdx)**<br /> Shows a configurable placeholder when the **ProseMirror** content of your editor is empty

**[StrikeExtension](./strike-extension.mdx)**<br /> Makes the text under the cursor / or at the provided position range ~~strike~~

**[SubExtension](./sub-extension.mdx)**<br /> Adds a `sub` mark to the editor. This is used to mark inline text as a <sub>subscript</sub> snippet

**[SupExtension](./sup-extension.mdx)**<br /> Adds a `sup` mark to the editor. This is used to mark inline text as a <sup>superscript</sup> snippet

**[TablesExtension](./tables-extension.mdx)**<br /> Adds simple tables to the editor

**[TextCaseExtension](./text-case-extension.mdx)**<br /> Formatting for text casing in the editor

**[TextColorExtension](./text-color-extension.mdx)**<br /> Makes the text under the cursor / or at the provided position range have the specified color

**[TextHighlightExtension](./text-highlight-extension.mdx)**<br /> Add a highlight color to the selected text

**[TrailingNodeExtension](./trailing-node-extension.mdx)**<br /> Ensure that there's always a trailing paragraph at the end of the document

**[UnderlineExtension](./underline-extension.mdx)**<br /> Makes the text under the cursor / or at the provided position range <u>underline</u>

**[WhitespaceExtension](./whitespace-extension.mdx)**<br /> Manage whitespace characters within the editor

**[YjsExtension](./yjs-extension.mdx)**<br /> The recommended extension for creating a collaborative editor, powered by `Yjs`.

## Additional extensions

The following extensions are _not_ bundled with the `remirror` package, and must be installed manually.

**[CountExtension](./count-extension.mdx)** <br /> Adds ability to count characters or words in your editor, and optionally highlight to a user when they have exceeded a specified limit.

**[FileExtension](./file-extension.mdx)** <br /> Adds the ability to upload arbitrary files to the editor.

## Community extensions

The following extensions are maintained by the community:

**[EntityExtension](https://github.com/Collaborne/remirror-entity-extension)**<br /> Stores all the data of an entity within the document itself

:::info

Did you create an extension and want to list it here as well? Please create a PR - or reach out on [Discord](https://remirror.io/chat).

:::
