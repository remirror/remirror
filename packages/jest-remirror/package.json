{
  "name": "jest-remirror",
  "version": "3.0.0-beta.6",
  "description": "A text editor for react built with prosemirror",
  "keywords": [
    "prosemirror",
    "react",
    "text editor"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/jest-remirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/jest-remirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/jest-remirror.js",
      "require": "./dist/jest-remirror.cjs"
    },
    "./environment": {
      "types": "./dist-types/environment.d.ts",
      "import": "./environment/dist/jest-remirror-environment.js",
      "require": "./environment/dist/jest-remirror-environment.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/jest-remirror.cjs",
  "module": "./dist/jest-remirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "environment"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core": "3.0.0-beta.6",
    "@remirror/dom": "3.0.0-beta.6",
    "@remirror/pm": "3.0.0-beta.4",
    "@remirror/preset-core": "3.0.0-beta.6",
    "@testing-library/dom": "^9.3.0",
    "@types/sanitize-html": "^2.9.0",
    "jest-prosemirror": "3.0.0-beta.4",
    "sanitize-html": "^2.10.0",
    "tiny-warning": "^1.0.3"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.0",
    "@types/node": "^16.18.34",
    "jest": "^29.5.0"
  },
  "peerDependencies": {
    "@types/node": "*",
    "jest": ">= 24.9.0"
  },
  "peerDependenciesMeta": {
    "@types/node": {
      "optional": true
    },
    "jest": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/jest-remirror.d.ts",
        "import": "./dist/jest-remirror.js",
        "require": "./dist/jest-remirror.cjs"
      },
      "./environment": {
        "types": "./environment/dist/jest-remirror-environment.d.ts",
        "import": "./environment/dist/jest-remirror-environment.js",
        "require": "./environment/dist/jest-remirror-environment.cjs"
      },
      "./package.json": "./package.json"
    },
    "main": "./dist/jest-remirror.cjs",
    "module": "./dist/jest-remirror.js",
    "type": "module",
    "types": "./dist/jest-remirror.d.ts"
  },
  "@remirror": {
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node",
            "jest"
          ]
        }
      }
    }
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "environment.ts"
    ]
  }
}
