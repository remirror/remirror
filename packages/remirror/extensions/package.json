{
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/remirror-extensions.d.ts",
      "import": "./dist/remirror-extensions.js",
      "require": "./dist/remirror-extensions.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/remirror-extensions.cjs",
  "module": "./dist/remirror-extensions.js",
  "types": "./dist/remirror-extensions.d.ts"
}
