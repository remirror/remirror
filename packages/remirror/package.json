{
  "name": "remirror",
  "version": "3.0.0-beta.8",
  "description": "One editing package to rule them all, one editing package to bind them.",
  "keywords": [
    "prosemirror",
    "react",
    "remirror",
    "wysiwyg",
    "editing",
    "text editor"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/remirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/remirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>",
    "ocavue "
  ],
  "sideEffects": false,
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/remirror.js",
      "require": "./dist/remirror.cjs"
    },
    "./dom": {
      "types": "./dist-types/dom.d.ts",
      "import": "./dom/dist/remirror-dom.js",
      "require": "./dom/dist/remirror-dom.cjs"
    },
    "./extensions": {
      "types": "./dist-types/extensions.d.ts",
      "import": "./extensions/dist/remirror-extensions.js",
      "require": "./extensions/dist/remirror-extensions.cjs"
    },
    "./icons": {
      "types": "./dist-types/icons.d.ts",
      "import": "./icons/dist/remirror-icons.js",
      "require": "./icons/dist/remirror-icons.cjs"
    },
    "./package.json": "./package.json",
    "./styles/*": "./styles/*"
  },
  "main": "./dist/remirror.cjs",
  "module": "./dist/remirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "dom",
    "extensions",
    "icons",
    "styles"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core": "3.0.0-beta.8",
    "@remirror/core-constants": "3.0.0-beta.4",
    "@remirror/core-helpers": "4.0.0-beta.5",
    "@remirror/core-types": "3.0.0-beta.5",
    "@remirror/core-utils": "3.0.0-beta.6",
    "@remirror/dom": "3.0.0-beta.8",
    "@remirror/extension-annotation": "3.0.0-beta.8",
    "@remirror/extension-bidi": "3.0.0-beta.8",
    "@remirror/extension-blockquote": "3.0.0-beta.8",
    "@remirror/extension-bold": "3.0.0-beta.8",
    "@remirror/extension-callout": "3.0.0-beta.8",
    "@remirror/extension-code": "3.0.0-beta.8",
    "@remirror/extension-code-block": "3.0.0-beta.8",
    "@remirror/extension-collaboration": "3.0.0-beta.8",
    "@remirror/extension-columns": "3.0.0-beta.8",
    "@remirror/extension-diff": "3.0.0-beta.8",
    "@remirror/extension-doc": "3.0.0-beta.8",
    "@remirror/extension-drop-cursor": "3.0.0-beta.8",
    "@remirror/extension-embed": "3.0.0-beta.8",
    "@remirror/extension-emoji": "3.0.0-beta.8",
    "@remirror/extension-entity-reference": "3.0.0-beta.8",
    "@remirror/extension-epic-mode": "3.0.0-beta.8",
    "@remirror/extension-events": "3.0.0-beta.8",
    "@remirror/extension-find": "1.0.0-beta.8",
    "@remirror/extension-font-family": "3.0.0-beta.8",
    "@remirror/extension-font-size": "3.0.0-beta.8",
    "@remirror/extension-gap-cursor": "3.0.0-beta.8",
    "@remirror/extension-hard-break": "3.0.0-beta.8",
    "@remirror/extension-heading": "3.0.0-beta.8",
    "@remirror/extension-history": "3.0.0-beta.8",
    "@remirror/extension-horizontal-rule": "3.0.0-beta.8",
    "@remirror/extension-image": "3.0.0-beta.8",
    "@remirror/extension-italic": "3.0.0-beta.8",
    "@remirror/extension-link": "3.0.0-beta.8",
    "@remirror/extension-list": "3.0.0-beta.8",
    "@remirror/extension-markdown": "3.0.0-beta.8",
    "@remirror/extension-mention": "3.0.0-beta.8",
    "@remirror/extension-mention-atom": "3.0.0-beta.8",
    "@remirror/extension-node-formatting": "3.0.0-beta.8",
    "@remirror/extension-paragraph": "3.0.0-beta.8",
    "@remirror/extension-placeholder": "3.0.0-beta.8",
    "@remirror/extension-positioner": "3.0.0-beta.8",
    "@remirror/extension-shortcuts": "3.0.0-beta.8",
    "@remirror/extension-strike": "3.0.0-beta.8",
    "@remirror/extension-sub": "3.0.0-beta.8",
    "@remirror/extension-sup": "3.0.0-beta.8",
    "@remirror/extension-tables": "3.0.0-beta.8",
    "@remirror/extension-text": "3.0.0-beta.8",
    "@remirror/extension-text-case": "3.0.0-beta.8",
    "@remirror/extension-text-color": "3.0.0-beta.8",
    "@remirror/extension-text-highlight": "3.0.0-beta.8",
    "@remirror/extension-trailing-node": "3.0.0-beta.8",
    "@remirror/extension-underline": "3.0.0-beta.8",
    "@remirror/extension-whitespace": "3.0.0-beta.8",
    "@remirror/icons": "3.0.0-beta.5",
    "@remirror/preset-core": "3.0.0-beta.8",
    "@remirror/preset-formatting": "3.0.0-beta.8",
    "@remirror/preset-wysiwyg": "3.0.0-beta.8",
    "@remirror/theme": "3.0.0-beta.5",
    "@types/refractor": "^3.0.2",
    "refractor": "^3.6.0"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.1",
    "@remirror/pm": "3.0.0-beta.6",
    "@types/prettier": "^2.7.2",
    "prettier": "^2.8.8"
  },
  "peerDependencies": {
    "@remirror/pm": "^3.0.0-beta.6",
    "@types/prettier": "^2.7.2",
    "prettier": "^2.8.8"
  },
  "peerDependenciesMeta": {
    "@types/prettier": {
      "optional": true
    },
    "prettier": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/remirror.d.ts",
        "import": "./dist/remirror.js",
        "require": "./dist/remirror.cjs"
      },
      "./dom": {
        "types": "./dom/dist/remirror-dom.d.ts",
        "import": "./dom/dist/remirror-dom.js",
        "require": "./dom/dist/remirror-dom.cjs"
      },
      "./extensions": {
        "types": "./extensions/dist/remirror-extensions.d.ts",
        "import": "./extensions/dist/remirror-extensions.js",
        "require": "./extensions/dist/remirror-extensions.cjs"
      },
      "./icons": {
        "types": "./icons/dist/remirror-icons.d.ts",
        "import": "./icons/dist/remirror-icons.js",
        "require": "./icons/dist/remirror-icons.cjs"
      },
      "./package.json": "./package.json",
      "./styles/*": "./styles/*"
    },
    "main": "./dist/remirror.cjs",
    "module": "./dist/remirror.js",
    "type": "module",
    "types": "./dist/remirror.d.ts"
  },
  "@remirror": {},
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "dom.ts",
      "extensions.ts",
      "icons.ts"
    ]
  }
}
