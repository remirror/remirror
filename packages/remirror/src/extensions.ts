// Extensions
export * from '@remirror/extension-annotation';
export * from '@remirror/extension-bidi';
export * from '@remirror/extension-blockquote';
export * from '@remirror/extension-bold';
export * from '@remirror/extension-callout';
export * from '@remirror/extension-code';
export * from '@remirror/extension-code-block';
export * from '@remirror/extension-codemirror5';
export * from '@remirror/extension-collaboration';
export * from '@remirror/extension-columns';
export * from '@remirror/extension-diff';
export * from '@remirror/extension-doc';
export * from '@remirror/extension-drop-cursor';
export * from '@remirror/extension-embed';
export * from '@remirror/extension-emoji';
export * from '@remirror/extension-entity-reference';
export * from '@remirror/extension-epic-mode';
export * from '@remirror/extension-events';
export * from '@remirror/extension-find';
export * from '@remirror/extension-font-family';
export * from '@remirror/extension-font-size';
export * from '@remirror/extension-gap-cursor';
export * from '@remirror/extension-hard-break';
export * from '@remirror/extension-heading';
export * from '@remirror/extension-history';
export * from '@remirror/extension-horizontal-rule';
export * from '@remirror/extension-image';
export * from '@remirror/extension-italic';
export * from '@remirror/extension-link';
export * from '@remirror/extension-list';
export * from '@remirror/extension-markdown';
export * from '@remirror/extension-mention';
export * from '@remirror/extension-mention-atom';
export * from '@remirror/extension-node-formatting';
export * from '@remirror/extension-paragraph';
export * from '@remirror/extension-placeholder';
export * from '@remirror/extension-positioner';
export * from '@remirror/extension-shortcuts';
export * from '@remirror/extension-strike';
export * from '@remirror/extension-sub';
export * from '@remirror/extension-sup';
export * from '@remirror/extension-tables';
export * from '@remirror/extension-text';
export * from '@remirror/extension-text-case';
export * from '@remirror/extension-text-color';
export * from '@remirror/extension-text-highlight';
export * from '@remirror/extension-trailing-node';
export * from '@remirror/extension-underline';
export * from '@remirror/extension-whitespace';
export * from '@remirror/extension-yjs';

// Presets
export * from '@remirror/preset-core';
export * from '@remirror/preset-formatting';
export * from '@remirror/preset-wysiwyg';
