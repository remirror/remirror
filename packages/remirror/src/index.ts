export * from '@remirror/core';
export * from '@remirror/theme';
