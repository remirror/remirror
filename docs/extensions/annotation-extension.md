---
hide_title: true
title: AnnotationExtension
---

# `AnnotationExtension`

## Summary

This extension allows to annotate the content in your editor

## Features

## Annotate multiple text nodes

Annotation enrich parts of a document. For example, a user could annotate a sentence as "important" to find it back later on.

Annotations are rendered as [Decorations](https://prosemirror.net/docs/ref/#view.Decorations), these differ from [marks](https://prosemirror.net/docs/ref/#model.Mark) in that they can span across multiple Prosemirror nodes.

Prosemirror stores all content in a [flat sequence of nodes](https://prosemirror.net/docs/guide/#doc). For example, the text "**bold _italic_ bold**" contains 3 different nodes (italic forces a split of the node)

| Text     | Marks            |
| -------- | ---------------- |
| `bold `  | `bold`           |
| `italic` | `bold`, `italic` |
| ` bold`  | `bold`           |

In contrast, one annotation could cover **_multiple_** text nodes with a **_single_** decoration. This is relevant e.g. if you want to show a list of sentences marked as "important" or if you want to allow users to rename the "important" annotation (which otherwise would have to be done in all nodes).

### Overlapping annotations

Annotations can be partially or fully overlapping. For example, a user could annotate the "important" sentence as well with "to be reviewed" or a word in the sentence as "customer X".

The annotation extension provides logic to visualize such overlapping annotations by mixing colors. See storybook for an example.

### Extendable data model

The extension defines only the minimal required fields: position where the annotation starts/ends and an ID. For convince, the annotation provides also the text covered by the annotation.

An app using the annotation-extension can extend the base data model. For example, it could add a label (like "important") or a color to each annotation. The annotation extension will pass these custom fields simply through to the app.

### Collaborative editing (Yjs)

As stated above, Annotations are _decorations_ - these are not part of the Prosemirror document model, they are part of the view. Whilst the model is syncronised between users, the _view_ is not.

To enable collaboration on annotations, additional logic has been added in the [Yjs extension](https://github.com/remirror/remirror/tree/main/packages/remirror__extension-yjs#readme), that modifies _this_ extension's options to utilise a [Yjs Map](https://docs.yjs.dev/api/shared-types/y.map). This is a shared data structure, meaning annotations can also be collaborated on.

This [proof of concept](https://github.com/collaborne/remirror-yjs-webrtc-demo/#readme) has a in-depth description of this approach.

## Usage

### Installation

This extension is installed for you when you install the main `remirror` package.

You can use the imports in the following way.

```ts
import { AnnotationExtension } from 'remirror/extensions';
```

To install it directly you can use

The extension is provided by the `@remirror/extension-annotation` package. There are two ways of pulling it into your project.

### Examples

See storybook for example usage.

## API

### Options

### Commands

### Helpers
