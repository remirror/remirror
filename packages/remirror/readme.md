# remirror

> One package to rule them all, one entry point to bind them.

[![Version][version]][npm] [![Weekly Downloads][downloads-badge]][npm] [![Bundled size][size-badge]][size] [![Typed Codebase][typescript]](#) [![MIT License][license]](#)

[version]: https://flat.badgen.net/npm/v/remirror
[npm]: https://npmjs.com/package/remirror
[license]: https://flat.badgen.net/badge/license/MIT/purple
[size]: https://bundlephobia.com/result?p=remirror
[size-badge]: https://flat.badgen.net/bundlephobia/minzip/remirror
[typescript]: https://flat.badgen.net/badge/icon/TypeScript?icon=typescript&label
[downloads-badge]: https://badgen.net/npm/dw/remirror/red?icon=npm

## Installation

```bash
# yarn
yarn add remirror @remirror/pm

# pnpm
pnpm add remirror @remirror/pm

# npm
npm install remirror @remirror/pm
```

The `remirror` package will automatically install the `@remirror/pm` package for you. You only need to install it yourself if you'd like to ensure consistent versions of the `prosemirror-*` libraries when importing from `@remirror/pm/state` instead of `prosemirror-state` or `@remirror/pm/model` instead of `prosemirror-model`.

## Usage

Rather than installing multiple scoped packages, the `remirror` package is a gateway to using all the goodness that remirror provides while minimising your bundle size.

The following creates a controlled editor with React.

```tsx
import React from 'react';
import { SocialPreset } from 'remirror/extensions';
import { Remirror, SocialEmojiComponent, useRemirror } from '@remirror/react';

const EditorWrapper = () => {
  const socialPreset = new SocialPreset();
  const { state, onChange } = useRemirror({ extensions: () => [...socialPreset()] });

  return (
    <Remirror state={state} onChange={onChange} manager={manager} autoRender={true}>
      <SocialEmojiComponent />
    </Remirror>
  );
};
```

These are the entry points available through the `remirror` package.

- `remirror` - All the core functionality available through `@remirror/core`.
- `remirror/extensions` - All the core extensions and presets made available through the main `remirror` repository. This doesn't include any framework specific extensions and presets.
- `remirror/dom` - The dom framework implementation of via `createDomEditor`.
