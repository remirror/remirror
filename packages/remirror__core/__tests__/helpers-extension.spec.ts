import { extensionValidityTest, renderEditor } from 'jest-remirror';
import { BoldExtension, HeadingExtension, ItalicExtension } from 'remirror/extensions';

import { HelpersExtension } from '../';

extensionValidityTest(HelpersExtension);

describe('active', () => {
  it('should recognize active nodes by attrs', () => {
    const {
      add,
      nodes: { p, doc },
      attributeNodes: { heading: h },
      active,
    } = renderEditor<HeadingExtension>([new HeadingExtension()]);

    add(doc((p('one'), h({ level: 2 })('tw<cursor>o'))));

    expect(active.heading({ level: 2 })).toBeTrue();
    expect(active.heading({ level: 1 })).toBeFalse();
  });
});

describe('helpers', () => {
  describe('`isSelectionEmpty`', () => {
    it('returns true if the selection is empty', () => {
      const { add, nodes, helpers } = renderEditor<never>([]);
      const { p, doc } = nodes;

      add(doc(p('on<cursor>e')));
      expect(helpers.isSelectionEmpty()).toBeTrue();
    });

    it('returns false if the selection is not empty', () => {
      const { add, nodes, helpers } = renderEditor<never>([]);
      const { p, doc } = nodes;

      add(doc(p('on<start>e<end>')));
      expect(helpers.isSelectionEmpty()).toBeFalse();
    });
  });

  describe('`isViewEditable`', () => {
    it('returns true if the view is editable', () => {
      const { helpers } = renderEditor<never>([]);
      expect(helpers.isViewEditable()).toBeTrue();
    });

    it('returns false if the view is not editable', () => {
      const { helpers } = renderEditor<never>([], { props: { editable: false } });
      expect(helpers.isViewEditable()).toBeFalse();
    });
  });
});

describe('commands.insertHtml', () => {
  it('can insert html', () => {
    const editor = renderEditor([
      new HeadingExtension(),
      new BoldExtension(),
      new ItalicExtension(),
    ]);
    const { doc, p } = editor.nodes;
    const { bold, italic } = editor.marks;
    const { heading } = editor.attributeNodes;
    const h1 = heading({ level: 1 });

    editor.add(doc(p('Content<cursor>')));

    editor.chain.insertHtml('<h1>This is a heading</h1>').selectText('end').run();
    expect(editor.state.doc).toEqualProsemirrorNode(doc(p('Content'), h1('This is a heading')));

    editor.commands.insertHtml('<p>A paragraph <em>with</em> <strong>formatting</strong></p>');
    expect(editor.state.doc).toEqualProsemirrorNode(
      doc(
        p('Content'),
        h1('This is a heading'),
        p('A paragraph ', italic('with'), ' ', bold('formatting')),
      ),
    );
  });

  it('can insert marks', () => {
    const editor = renderEditor<BoldExtension>([new BoldExtension()]);
    const { doc, p } = editor.nodes;
    const { bold } = editor.marks;

    editor.add(doc(p('Content <cursor>')));

    editor.chain.insertHtml('<strong>is bold.</strong>').selectText('end').run();
    expect(editor.state.doc).toEqualProsemirrorNode(doc(p('Content ', bold('is bold.'))));
    expect(editor.dom).toMatchSnapshot();
  });
});
