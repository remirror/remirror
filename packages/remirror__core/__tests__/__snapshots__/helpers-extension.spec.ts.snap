// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands.insertHtml can insert marks 1`] = `
<div
  aria-label=""
  aria-multiline="true"
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
  translate="no"
>
  <p>
    Content 
    <strong>
      is bold.
    </strong>
  </p>
</div>
`;
